src/main/scala/com/itechart/ny_accidents/service/WeatherMappingService.scala

```diff
@@ -1,52 +1,80 @@
   // map under have such structure -> Map[stationId, Map[TimeHash, Seq[WeatherEntity]]]
-  private val allWeather: Map[Int, Map[Long, Seq[WeatherEntity]]] = Await.result(
-    NYDataDatabase.database.run(weatherDAO.allWeather()), Duration.Inf)
-    .sortBy(weather => weather.dateTime)
-    .groupBy(weather => weather.stationId)
-    .mapValues(weathers => {
-      weathers.map(weather => (DateUtils.hashByDate(weather.dateTime), weather))
-        .groupBy(_._1)
-        .mapValues(seq => seq.map(_._2))
+  private val allWeather: Map[Int, Map[Long, Seq[WeatherEntity]]] =
+    Await.result(
+      NYDataDatabase.database.run(weatherDAO.allWeather()), Duration.Inf)
+      .sortBy(weather => weather.dateTime)
+      .groupBy(weather => weather.stationId)
+      .mapValues(weathers => {
+        weathers.map(weather => (DateUtils.hashByDate(weather.dateTime), weather))
+          .groupBy(_._1)
+          .mapValues(seq => seq.map(_._2))
+      })
+
+  private val newSparkWeather: Array[((Int, Long), WeatherForAccident)] = weatherDAO.loadAllWeatherSpark()
+    .map(row => ((row.getInt(1), DateUtils.hashByDate(row.getLong(2))), DateUtils.weatherForAccidentMapper(row)))
+    .collect()
+    .sortWith((a, b) => {
+      if (a._1._1 < b._1._1) {
+        true
+      } else if (a._1._1 > b._1._1) {
+        false
+      } else a._1._2 < b._1._2
     })
+  private val minDate = newSparkWeather(0)._1._2
+  private val maxDate = newSparkWeather(newSparkWeather.length - 1)._1._2
 
+  private val weatherMap: Map[(Int, Long), WeatherForAccident] = newSparkWeather.toMap
 
-  def findWeatherByTimeAndCoordinates(accidentTime: Long, lat: Double, lon: Double): Option[WeatherForAccident] = {
-    val stationId = getNearestStationId(lat, lon)
-    // we take weather for current hour, for future one and for previous, because if we have got 10:58 accident and
-    // weather time for 10 o'clock is 10:05, and the next weather is 11:05, it's will be more accurately to take 11:05 weather
-    Seq(
-      Try(allWeather(stationId)(DateUtils.hashByDate(accidentTime))).toOption,
-      Try(allWeather(stationId)(DateUtils.hashByDate(DateUtils.addHour(accidentTime)))).toOption,
-      Try(allWeather(stationId)(DateUtils.hashByDate(DateUtils.subtractHour(accidentTime)))).toOption
-    ).filter(_.isDefined).map(_.get).reduceOption(_ ++ _) match {
-      case Some(value) => Some(findBestMatchWeather(value, accidentTime))
-      case _ => None
+  val weather: Dataset[((Int, Long), WeatherForAccident)] = Spark.sparkSql.createDataset((1 to 3)
+    .flatMap(stationId => {
+      (minDate to maxDate by HASH_DIFFERENCE)
+        .map(hash => {
+          ((stationId, hash), findBestWeather(stationId, hash, usePlus = false, 1))
+        })
+    }))
+
+  @tailrec private def findBestWeather(stationId: Int, dateHash: Long, usePlus: Boolean, currentDepth: Int): WeatherForAccident = {
+    if (currentDepth > 5)
+      return null
```

Comment:
never return Nulls in Scala. Use  Option instead

Reply:
https://github.com/itechartbigdatalab/NY-Accidents-Rate/pull/56/commits/0ca555bd973fed5eb36e11b9215abe2ce98024e1